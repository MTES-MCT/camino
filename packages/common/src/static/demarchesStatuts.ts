import { Definition } from '../definition.js'
import { Couleur } from './couleurs.js'

export const DemarchesStatutsIds = {
  Accepte: 'acc',
  ClasseSansSuite: 'cls',
  Depose: 'dep',
  Desiste: 'des',
  EnConstruction: 'eco',
  FinPoliceMines: 'fpm',
  Indetermine: 'ind',
  Initie: 'ini',
  EnInstruction: 'ins',
  Rejete: 'rej',
  Termine: 'ter'
} as const

export type DemarcheStatutId = typeof DemarchesStatutsIds[keyof typeof DemarchesStatutsIds]

export type DemarcheStatut<T = DemarcheStatutId> = Definition<T> & { couleur: Couleur }

export const DemarchesStatuts: {
  [key in DemarcheStatutId]: DemarcheStatut<key>
} = {
  acc: {
    id: 'acc',
    nom: 'accepté',
    description: "La demande a fait l’objet d’une décision favorable de l'administration.\n",
    couleur: 'success',
    ordre: 6
  },
  cls: {
    id: 'cls',
    nom: 'classé sans suite',
    description: 'La demande est classée sans suite par l’administration à défaut de pouvoir mener l’instruction à son terme.\n',
    couleur: 'neutral',
    ordre: 5
  },
  dep: {
    id: 'dep',
    nom: 'déposé',
    description: 'La demande est déposée auprès du service instructeur.\n',
    couleur: 'warning',
    ordre: 2
  },
  des: {
    id: 'des',
    nom: 'désisté',
    description:
      'La demande dont le porteur se dessaisit en informant l’administration de son choix de ne pas lui donner suite. Cette démarche demeure publique si elle a fait l’objet d’une consultation du public.',
    couleur: 'neutral',
    ordre: 8
  },
  eco: {
    id: 'eco',
    nom: 'en construction',
    description: 'Le dossier de demande est en cours d’élaboration par le demandeur.',
    couleur: 'warning',
    ordre: 1
  },
  fpm: {
    id: 'fpm',
    nom: 'fin de la police des mines',
    description: 'Fin de la police des mines.',
    couleur: 'success',
    ordre: 11
  },
  ind: {
    id: 'ind',
    nom: 'indéterminé',
    description: 'Les informations contenues dans les étapes de la démarche sont insuffisantes pour en déterminer le statut.\n',
    couleur: 'warning',
    ordre: 10
  },
  ini: {
    id: 'ini',
    nom: 'initié',
    description: 'La démarche est engagée unilatéralement par l’administration, elle ne fait pas l’objet d’une demande de la part du titulaire du titre ou de l’autorisation.',
    couleur: 'warning',
    ordre: 3
  },
  ins: {
    id: 'ins',
    nom: 'en instruction',
    description: 'La demande a fait l’objet d’un examen de complétude et de recevabilité.\n',
    couleur: 'warning',
    ordre: 4
  },
  rej: {
    id: 'rej',
    nom: 'rejeté',
    description:
      "La demande a fait l’objet d’une décision défavorable de l'administration. Les textes d’applications du code minier prévoient que les décisions de rejet ne font pas l’objet d’une publication. En conséquence, les démarches qui passent au statut “rejeté” dans Camino sont dé-publiées et rendues inaccessibles aux tiers.\n",
    couleur: 'error',
    ordre: 7
  },
  ter: {
    id: 'ter',
    nom: 'terminé',
    description:
      'La démarche engagée unilatéralement par l’administration qui a fait l’objet d’une décision de l’administration. Celle-ci n’a pas fait l’objet d’une demande de la part du titulaire du titre ou de l’autorisation.\n',
    couleur: 'success',
    ordre: 9
  }
} as const

export const sortedDemarchesStatuts = Object.values(DemarchesStatuts).sort((a, b) => a.ordre - b.ordre)
