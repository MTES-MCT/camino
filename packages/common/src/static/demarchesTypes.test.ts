import { sortedDemarchesStatuts } from './demarchesStatuts.js'
import { DemarchesTypesIds, DEMARCHES_TYPES_IDS, isDemarcheTypeId, isDemarcheTypeOctroi, isDemarcheTypeWithPhase, canImpactTitre } from './demarchesTypes.js'
import { test, expect } from 'vitest'

test('isDemarcheTypeId', () => {
  expect(isDemarcheTypeId(null)).toBe(false)
  expect(isDemarcheTypeId(undefined)).toBe(false)
  for (const demarcheType of DemarchesTypesIds) {
    expect(isDemarcheTypeId(demarcheType)).toBe(true)
  }
})

test('isDemarcheTypeOctroi', () => {
  for (const demarcheType of DemarchesTypesIds) {
    expect(isDemarcheTypeOctroi(demarcheType)).toBe(
      [DEMARCHES_TYPES_IDS.Octroi, DEMARCHES_TYPES_IDS.MutationPartielle, DEMARCHES_TYPES_IDS.Fusion, DEMARCHES_TYPES_IDS.DemandeDeTitreDExploitation].includes(demarcheType)
    )
  }
})

test('isDemarcheTypeWithPhase', () => {
  for (const demarcheType of DemarchesTypesIds) {
    let expected = isDemarcheTypeOctroi(demarcheType)

    if (!expected) {
      expected = [DEMARCHES_TYPES_IDS.Prolongation, DEMARCHES_TYPES_IDS.Prolongation1, DEMARCHES_TYPES_IDS.Prolongation2, DEMARCHES_TYPES_IDS.ProlongationExceptionnelle].includes(demarcheType)
    }

    expect(isDemarcheTypeWithPhase(demarcheType)).toBe(expected)
  }
})

test('canImpactTitre', () => {
  const result = DemarchesTypesIds.flatMap(demarcheType => {
    return sortedDemarchesStatuts.map(demarcheStatus => {
      return `${demarcheType} - ${demarcheStatus.id} -> ${canImpactTitre(demarcheType, demarcheStatus.id)}`
    })
  })
  expect(result).toMatchSnapshot()
})
