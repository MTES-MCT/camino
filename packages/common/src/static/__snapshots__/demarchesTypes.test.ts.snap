// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`canImpactTitre 1`] = `
[
  "amo - acc -> true",
  "amo - acp -> true",
  "amo - cls -> false",
  "amo - dep -> false",
  "amo - des -> false",
  "amo - eco -> false",
  "amo - ins -> false",
  "amo - fpm -> false",
  "amo - ind -> false",
  "amo - ini -> false",
  "amo - rej -> false",
  "amo - rea -> false",
  "amo - ter -> true",
  "aom - acc -> true",
  "aom - acp -> true",
  "aom - cls -> false",
  "aom - dep -> false",
  "aom - des -> false",
  "aom - eco -> false",
  "aom - ins -> false",
  "aom - fpm -> false",
  "aom - ind -> false",
  "aom - ini -> false",
  "aom - rej -> false",
  "aom - rea -> false",
  "aom - ter -> true",
  "ces - acc -> true",
  "ces - acp -> true",
  "ces - cls -> false",
  "ces - dep -> false",
  "ces - des -> false",
  "ces - eco -> false",
  "ces - ins -> false",
  "ces - fpm -> false",
  "ces - ind -> false",
  "ces - ini -> false",
  "ces - rej -> false",
  "ces - rea -> false",
  "ces - ter -> true",
  "con - acc -> true",
  "con - acp -> true",
  "con - cls -> false",
  "con - dep -> false",
  "con - des -> false",
  "con - eco -> false",
  "con - ins -> false",
  "con - fpm -> false",
  "con - ind -> false",
  "con - ini -> false",
  "con - rej -> false",
  "con - rea -> false",
  "con - ter -> true",
  "dam - acc -> true",
  "dam - acp -> true",
  "dam - cls -> false",
  "dam - dep -> false",
  "dam - des -> false",
  "dam - eco -> false",
  "dam - ins -> false",
  "dam - fpm -> false",
  "dam - ind -> false",
  "dam - ini -> false",
  "dam - rej -> false",
  "dam - rea -> false",
  "dam - ter -> true",
  "dec - acc -> true",
  "dec - acp -> true",
  "dec - cls -> false",
  "dec - dep -> false",
  "dec - des -> false",
  "dec - eco -> false",
  "dec - ins -> false",
  "dec - fpm -> false",
  "dec - ind -> false",
  "dec - ini -> false",
  "dec - rej -> false",
  "dec - rea -> false",
  "dec - ter -> true",
  "dep - acc -> true",
  "dep - acp -> true",
  "dep - cls -> false",
  "dep - dep -> false",
  "dep - des -> false",
  "dep - eco -> false",
  "dep - ins -> false",
  "dep - fpm -> false",
  "dep - ind -> false",
  "dep - ini -> false",
  "dep - rej -> false",
  "dep - rea -> false",
  "dep - ter -> true",
  "dot - acc -> true",
  "dot - acp -> true",
  "dot - cls -> false",
  "dot - dep -> false",
  "dot - des -> false",
  "dot - eco -> false",
  "dot - ins -> false",
  "dot - fpm -> false",
  "dot - ind -> false",
  "dot - ini -> false",
  "dot - rej -> false",
  "dot - rea -> false",
  "dot - ter -> true",
  "exp - acc -> true",
  "exp - acp -> true",
  "exp - cls -> false",
  "exp - dep -> false",
  "exp - des -> false",
  "exp - eco -> false",
  "exp - ins -> false",
  "exp - fpm -> false",
  "exp - ind -> false",
  "exp - ini -> false",
  "exp - rej -> false",
  "exp - rea -> false",
  "exp - ter -> true",
  "exs - acc -> true",
  "exs - acp -> true",
  "exs - cls -> false",
  "exs - dep -> false",
  "exs - des -> false",
  "exs - eco -> false",
  "exs - ins -> false",
  "exs - fpm -> false",
  "exs - ind -> false",
  "exs - ini -> false",
  "exs - rej -> false",
  "exs - rea -> false",
  "exs - ter -> true",
  "fus - acc -> true",
  "fus - acp -> true",
  "fus - cls -> true",
  "fus - dep -> true",
  "fus - des -> true",
  "fus - eco -> true",
  "fus - ins -> true",
  "fus - fpm -> true",
  "fus - ind -> true",
  "fus - ini -> true",
  "fus - rej -> true",
  "fus - rea -> true",
  "fus - ter -> true",
  "mut - acc -> true",
  "mut - acp -> true",
  "mut - cls -> false",
  "mut - dep -> false",
  "mut - des -> false",
  "mut - eco -> false",
  "mut - ins -> false",
  "mut - fpm -> false",
  "mut - ind -> false",
  "mut - ini -> false",
  "mut - rej -> false",
  "mut - rea -> false",
  "mut - ter -> true",
  "oct - acc -> true",
  "oct - acp -> true",
  "oct - cls -> true",
  "oct - dep -> true",
  "oct - des -> true",
  "oct - eco -> true",
  "oct - ins -> true",
  "oct - fpm -> true",
  "oct - ind -> true",
  "oct - ini -> true",
  "oct - rej -> true",
  "oct - rea -> true",
  "oct - ter -> true",
  "pr1 - acc -> true",
  "pr1 - acp -> true",
  "pr1 - cls -> false",
  "pr1 - dep -> false",
  "pr1 - des -> false",
  "pr1 - eco -> false",
  "pr1 - ins -> false",
  "pr1 - fpm -> false",
  "pr1 - ind -> false",
  "pr1 - ini -> false",
  "pr1 - rej -> false",
  "pr1 - rea -> false",
  "pr1 - ter -> true",
  "pr2 - acc -> true",
  "pr2 - acp -> true",
  "pr2 - cls -> false",
  "pr2 - dep -> false",
  "pr2 - des -> false",
  "pr2 - eco -> false",
  "pr2 - ins -> false",
  "pr2 - fpm -> false",
  "pr2 - ind -> false",
  "pr2 - ini -> false",
  "pr2 - rej -> false",
  "pr2 - rea -> false",
  "pr2 - ter -> true",
  "pre - acc -> true",
  "pre - acp -> true",
  "pre - cls -> false",
  "pre - dep -> false",
  "pre - des -> false",
  "pre - eco -> false",
  "pre - ins -> false",
  "pre - fpm -> false",
  "pre - ind -> false",
  "pre - ini -> false",
  "pre - rej -> false",
  "pre - rea -> false",
  "pre - ter -> true",
  "pro - acc -> true",
  "pro - acp -> true",
  "pro - cls -> false",
  "pro - dep -> false",
  "pro - des -> false",
  "pro - eco -> false",
  "pro - ins -> false",
  "pro - fpm -> false",
  "pro - ind -> false",
  "pro - ini -> false",
  "pro - rej -> false",
  "pro - rea -> false",
  "pro - ter -> true",
  "prr - acc -> true",
  "prr - acp -> true",
  "prr - cls -> false",
  "prr - dep -> false",
  "prr - des -> false",
  "prr - eco -> false",
  "prr - ins -> false",
  "prr - fpm -> false",
  "prr - ind -> false",
  "prr - ini -> false",
  "prr - rej -> false",
  "prr - rea -> false",
  "prr - ter -> true",
  "ren - acc -> true",
  "ren - acp -> true",
  "ren - cls -> false",
  "ren - dep -> false",
  "ren - des -> false",
  "ren - eco -> false",
  "ren - ins -> false",
  "ren - fpm -> false",
  "ren - ind -> false",
  "ren - ini -> false",
  "ren - rej -> false",
  "ren - rea -> false",
  "ren - ter -> true",
  "res - acc -> true",
  "res - acp -> true",
  "res - cls -> false",
  "res - dep -> false",
  "res - des -> false",
  "res - eco -> false",
  "res - ins -> false",
  "res - fpm -> false",
  "res - ind -> false",
  "res - ini -> false",
  "res - rej -> false",
  "res - rea -> false",
  "res - ter -> true",
  "ret - acc -> true",
  "ret - acp -> true",
  "ret - cls -> false",
  "ret - dep -> false",
  "ret - des -> false",
  "ret - eco -> false",
  "ret - ins -> false",
  "ret - fpm -> false",
  "ret - ind -> false",
  "ret - ini -> false",
  "ret - rej -> false",
  "ret - rea -> false",
  "ret - ter -> true",
  "vct - acc -> true",
  "vct - acp -> true",
  "vct - cls -> true",
  "vct - dep -> true",
  "vct - des -> true",
  "vct - eco -> true",
  "vct - ins -> true",
  "vct - fpm -> true",
  "vct - ind -> true",
  "vct - ini -> true",
  "vct - rej -> true",
  "vct - rea -> true",
  "vct - ter -> true",
  "vut - acc -> true",
  "vut - acp -> true",
  "vut - cls -> true",
  "vut - dep -> true",
  "vut - des -> true",
  "vut - eco -> true",
  "vut - ins -> true",
  "vut - fpm -> true",
  "vut - ind -> true",
  "vut - ini -> true",
  "vut - rej -> true",
  "vut - rea -> true",
  "vut - ter -> true",
]
`;

exports[`isDemarcheTypeProlongations 1`] = `
[
  "amo -> false",
  "aom -> false",
  "ces -> false",
  "con -> false",
  "dam -> false",
  "dec -> false",
  "dep -> false",
  "dot -> false",
  "exp -> false",
  "exs -> false",
  "fus -> false",
  "mut -> false",
  "oct -> false",
  "pr1 -> true",
  "pr2 -> true",
  "pre -> true",
  "pro -> true",
  "prr -> false",
  "ren -> false",
  "res -> false",
  "ret -> false",
  "vct -> false",
  "vut -> false",
]
`;
